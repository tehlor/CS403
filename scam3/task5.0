(define ones (cons-stream 1 ones))
(sdisplay 12 (smush ones +))
(println)