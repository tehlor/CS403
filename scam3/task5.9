(include "random.lib")
(randomSeed 116)
(define (rs) (cons-stream (randomRange -99 100) (rs)))
(define s (rs))
(define op (getElement (list + - *) (randomRange 0 3)))
(inspect op)
(define t (smush s op))
(sdisplay 12 s) (println)
(sdisplay 12 t) (println)
