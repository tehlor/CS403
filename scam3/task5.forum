(include "random.lib")
(randomSeed 36)
(define s (cons-stream (randomRange -10 10) s))
(define t (smush s (getElement (list + - * /) (randomRange 0 4))))
(sdisplay 12 t) (println)