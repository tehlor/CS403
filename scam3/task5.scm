(define (main)
    (setPort (open (getElement ScamArgs 1) 'read))
    (setNilDisplay 'nil)
    (readArgs)
)

(define (readArgs)
    (define env this)
    (define (iter expr)
        (if (not (eof?))
            (begin (eval expr env) 
                (iter (readExpr))
            )
        )
    )
    (iter (readExpr))
)

(define (smush S operator)
    (make S operator 0 0 nil)
)

(define (make S operator i mode result)
    (cond 
        ((== mode 0)
            (cons-stream i (make S operator (+ i 1) 1 result))
        )
        ((== mode 1)
            (cons-stream (stream-car S) (make S operator i 2 result))
        )
        ((== mode 2)
            (cond 
                ((== result nil)
                    (cons-stream (stream-car S) (make (stream-cdr S) operator i 0 (stream-car S)))
                )
                (else
                    (define r (operator result (stream-car S)))
                    (cons-stream r (make (stream-cdr S) operator i 0 r))
                )
            )
        )
    )
)

(define (sdisplay num s)
    (define (iter i S)
        (cond 
            ((< i num)
                (print (stream-car S) ",")
                (iter (+ i 1) (stream-cdr S))
            )
            (else
                (print "...")
            )
        )
    )
    (print "(")
    (iter 0 s)
    (print ")")
)
