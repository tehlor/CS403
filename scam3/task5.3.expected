op is <builtin +(@)>
(-89,-14,-19,-43,-31,38,-1,16,42,99,-93,87,...)
(0,-89,-89,1,-14,-103,2,-19,-122,3,-43,-165,...)